controllers/machinepool_controller_noderef.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	apicorev1 "k8s.io/api/core/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/controllers/noderefutil"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	capierrors "sigs.k8s.io/cluster-api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type getNodeReferencesResult struct {
+	references []apicorev1.ObjectReference
+	available  int
+	ready      int
+}
+
+func (r *MachinePoolReconciler) reconcileNodeRefs(ctx context.Context, cluster *clusterv1.Cluster, machinepool *clusterv1.MachinePool) error {
+	logger := r.Log.WithValues("machinepool", machinepool.Name, "namespace", machinepool.Namespace)
+	// Check that the MachinePool hasn't been deleted or in the process.
+	if !machinepool.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// Check that the Machine doesn't already have a NodeRefs.
+	if len(machinepool.Status.NodeRefs) == int(machinepool.Status.Replicas) {
```

Comment:
What if we add and remove a replica very quickly? Would that get us to have stale data?

Reply:
I changed this to be a bit more robust. Now it also checks if replicas == readyReplicas, with readyReplicas being reset if the infraProvider state changes.